index.bs

```diff
@@ -4046,7 +4073,63 @@ To override this default policy and indicate that a cross-origin <{iframe}> is a
 
 [=[RPS]=] utilizing the WebAuthn API in an embedded context should review [[#sctn-seccons-visibility]] regarding [=UI redressing=] and its possible mitigations.
 
+## Using Web Authentication across related origins ## {#sctn-related-origins}
+
+By default, Web Authentication requires that the [=RP ID=] be equal to the [=determines the set of origins on which the public key credential may be exercised|origin=]'s [=effective domain=], or a [=is a registrable domain suffix of or is equal to|registrable domain suffix=] of the [=determines the set of origins on which the public key credential may be exercised|origin=]'s [=effective domain=].
+
+This can make deployment challenging for large environments where multiple country-specific domains are in use (e.g. example.com vs example.co.uk vs example.sg), where alternative or brand domains are required (e.g. myexampletravel.com vs examplecruises.com), and/or where platform as a service providers are used to support mobile apps. 
+
+[=[WRPS]=] can opt in to allowing [=WebAuthn Clients=] to enable a credential to be created and used across a limited set of related [=origin|origins=].
+Such [=[RPS]=] MUST choose a common [=RP ID=] to use across all ceremonies from related origins.
+
+A JSON document MUST be hosted at the `webauthn` well-known URL ([[!RFC8615]]) for the [=RP ID=]. The JSON document MUST be returned as follows:
 
+    - The content type MUST be `application/json`.
+    - The top-level JSON object MUST contain a key named `origins` whose value MUST be an array of one or more strings containing web origins.
+
+For example, for the RP ID `example.com`:
+
+<xmp class="json" highlight="json">
+{
+    "origins": [
+        "https://example.co.uk",
+        "https://example.de",
+        "https://example.sg",
+        "https://example.net",
+        "https://exampledelivery.com",
+        "https://exampledelivery.co.uk",
+        "https://exampledelivery.de",
+        "https://exampledelivery.sg",
+        "https://myexamplerewards.com",
+        "https://examplecars.com"
+    ]
+}
+</xmp>
+
+[=WebAuthn Clients=] supporting this feature MUST support at least five [=origin labels=]. Client policy SHOULD define an upper limit to prevent abuse.
+
+[=WebAuthn Clients=] supporting this feature SHOULD include {{ClientCapability/relatedOrigins}} in their response to [[#sctn-getClientCapabilities|getClientCapabilities()]].
+
+### Validating Related Origins ### {#sctn-validating-relation-origin}
+
+To validate the calling origin is an authorized related origin for a given ceremony:
```

Comment:
This should explicitly define _callerOrigin_ and _rpIdRequested_ as parameters of the procedure, and the references in the client algorithms should explicitly specify the arguments.

I think we should also put a `<dfn>` here and reference that rather than simply linking to the section with a different link text. Right now the link text "related origins validation procedure" appears nowhere else in the document, so readers have to rely solely on the hyperlink in order to discover what is actually referenced (this is not the only case of that throughout the spec, but IMO it's preferable to minimize them).

Reply:
> This should explicitly define callerOrigin and rpIdRequested as parameters of the procedure, and the references in the client algorithms should explicitly specify the arguments.

Do you have an example of what that should look like in Bikeshed?

> I think we should also put a <dfn> here and reference that rather than simply linking to the section with a different link text. Right now the link text "related origins validation procedure" appears nowhere else in the document, so readers have to rely solely on the hyperlink in order to discover what is actually referenced (this is not the only case of that throughout the spec, but IMO it's preferable to minimize them).

Not sure I understand. What would a definition look like for a procedure? Isn't that just the section that defines the procedue?

Reply:
We have a few examples in the spec:

- [**determines the set of origins on which the public key credential may be exercised**](https://w3c.github.io/webauthn/#determines-the-set-of-origins-on-which-the-public-key-credential-may-be-exercised) ([source](https://github.com/w3c/webauthn/blob/bb9f9bb53c17e50f20a5ab31e754f1f644362fb1/index.bs#L1340))
- [**looking up**](https://w3c.github.io/webauthn/#credential-id-looking-up) a credential ID in an authenticator ([source](https://github.com/w3c/webauthn/blob/bb9f9bb53c17e50f20a5ab31e754f1f644362fb1/index.bs#L4644))
- [**Create a new supplemental public key record**](https://w3c.github.io/webauthn/#abstract-opdef-create-a-new-supplemental-public-key-record) ([source](https://github.com/w3c/webauthn/blob/bb9f9bb53c17e50f20a5ab31e754f1f644362fb1/index.bs#L7582))
- (Arguably) [**effective resident key requirement for credential creation**](https://w3c.github.io/webauthn/#effective-resident-key-requirement-for-credential-creation) ([source](https://github.com/w3c/webauthn/blob/bb9f9bb53c17e50f20a5ab31e754f1f644362fb1/index.bs#L1998))

So for example, this section could start with:

```
### Validating Related Origins ### {#sctn-validating-relation-origin}

The <dfn abstract-op>related origins validation procedure</dfn>, given arguments |callerOrigin| and |rpIdRequested|,
is as follows:
```

and then the references in `create()` and `get()` could look like this:
```
                            1. Run the [$related origins validation procedure$] with arguments |callerOrigin| and |rpIdRequested|.
                                If the result is [FALSE], throw a "{{SecurityError}}" {{DOMException}}.
```

This has a few advantages: most importantly it makes the connection clearer and less reliant on hidden metadata, so it's more likely to survive format conversions such as into PDF or plain text (or even actual print, if someone were to actually do that), or quotations. Less importantly it also shows Bikeshed's list of references when you click the definition, which doesn't happen for links to sections.

Alternatively, the link to the section should be a plain section reference without replacing the link text, so that it renders the whole section heading, for the reasons above.

Like I said, we don't always respect these concerns throughout the spec, and I don't think we _always_ should, but I think this is one of the cases where there's little or nothing to gain from not doing so.

Reply:
But to be fair, there is also at least this counter-example where we do link to a section with a (slightly) different link text:

- [limited verification algorithm](https://w3c.github.io/webauthn/#clientdatajson-development) ([source](https://github.com/w3c/webauthn/blob/bb9f9bb53c17e50f20a5ab31e754f1f644362fb1/index.bs#L3836))

Reply:
Makes total sense. Addressed in [9da2c4b](https://github.com/w3c/webauthn/pull/2040/commits/9da2c4b4f55291376bf8da81ac9a4479e39c029b). Thanks for explaining @emlun!